### PR TITLE
mesh: store bound reference pointers on a ComputedRoutes resource and use during reconcile

### PR DESCRIPTION
**Pull request by rboyer**
### Description

xRoute resource types contain a slice of parentRefs to services that they manipulate traffic for. All xRoutes that have a parentRef to given Service will be merged together to generate a ComputedRoutes resource name-aligned with that Service.

This means that a write of an xRoute with 2 parent ref pointers will cause at most 2 reconciles for ComputedRoutes.

If that xRoute's list of parentRefs were ever to be reduced, or otherwise lose an item, that subsequent map event will only emit events for **the current set of refs**. The removed ref will not cause the generated ComputedRoutes related to that service to be re-reconciled to omit the influence of that xRoute.

To combat this, we will store on the ComputedRoutes resource a `BoundResources []*pbresource.Reference` field with references to all resources that were used to influence the generated output.

When the routes controller reconciles, it will use a `bimapper` to index this influence, and the dependency mappers for the xRoutes will look themselves up in that index to discover additional (former) ComputedRoutes that need to be notified as well.

NET-5684

### CODE REVIEW

File: internal/mesh/internal/controllers/routes/xroutemapper/xroutemapper.go [hunk 1]
```diff
@@ -88,6 +92,17 @@ func (m *Mapper) walkRouteBackendBiMappers(fn func(bm *bimapper.Mapper)) {
 	}
 }
 
+func (m *Mapper) TrackComputedRoutes(cr *types.DecodedComputedRoutes) {
+	if cr != nil {
+		refs := refSliceToRefSlice(cr.Data.BoundReferences)
```
Comment (rboyer) [inline internal/mesh/internal/controllers/routes/xroutemapper/xroutemapper.go:97]: This is a mixed types index, which is new.

File: internal/mesh/internal/controllers/routes/generate.go [hunk 2]
```diff
@@ -68,6 +70,8 @@ func compile(
 	}
 	parentServiceID = parentServiceDec.Resource.Id // get ULID out of it
 
+	boundRefCollector.AddRefOrID(parentServiceRef)
```
Comment (rboyer) [inline internal/mesh/internal/controllers/routes/generate.go:73]: I only add it to the collector if the following are all true:
- we load the resource for some reason
- the resource is **found**
- we decided to use the information in that resource to produce `ComputedRoutes`

Those seem like suitable criteria to then use the persisted bound references to then later purge influence correctly.
-

Comment (rboyer) [reply internal/mesh/internal/controllers/routes/generate.go:73]: I decided to add all of the relevant references to the generated resource, if only for diagnostic reasons.

I _think_ only the xRoute mappers need to be updated due to their unique indexing strategy (meaningless name, ref in body).

**Review by ishustava — APPROVED** [review]
Looks good! A couple of minor comments

**Pull request merged**